### ISSUE TITLE
Support for default-to-TLS

### ISSUE THREAD

**Issue by cduchesne** [codedellemc/rexray#583]
Currently REX-Ray communicates to a remote REX-Ray server via the clear-text http protocol. We should enable server-side SSL and use https protocol by default.

**Comment by akutz** [comment]
This exists but isn't enabled by default because shipping with default HTTPS using certificates included with the product is no more secure than plain-text.

**Comment by akutz** [comment]
Documentation for libStorage [client and server TLS](http://libstorage.readthedocs.io/en/stable/user-guide/config/#tcptls).

**Comment by cduchesne** [comment]
Great - this can track the inclusion into REX-Ray as part of deployment. We shouldn't have to ship with certs, as we could potentially generate them during implementation as an option.

https://golang.org/src/crypto/tls/generate_cert.go

**Comment by akutz** [comment]
Hi @cduchesne,

Yeah, I know how to generate a certificate :) 

This isn't a minor issue though, IMO, in terms of work effort. If we start towards the business of generating certificates then what we should offer is the ability for a libStorage server's API to generate a cert pair that can be used as a client certificate for a libStorage client node. The pair would be signed by the libStorage server cert so that client-cert requirement is easily enabled.

Generating a server cert isn't enough IMO because the value is requiring the client-cert, and of course that has to be signed by the same CA (intermediate or otherwise) or by the server cert itself to work. This wouldn't happen automatically during deployment of the libStorage server alone, hence my above comment.

**Comment by cduchesne** [comment]
This is why there are 2 separate issues. Server-side SSL is important to encrypt communication and client-side SSL is important to authenticate clients. This feature is less important than the client-side certificate implementation, though it will check the box that states communication traffic passed is not in clear-text.

**Comment by akutz** [comment]
It's the same issue -- generating a certificate  chain about which both sides have knowledge. You cannot encrypt either way unless the remote side accepts the originating cert as valid. Not unless you disable verification, and then again, what is the point.

Also, the way client cert verification works is the server determines that the presented client cert is signed by the same CA as the server very (or by the server cert). This is why an intermediate CA is often used.

I've been down this road many times. I understand the issue and the requirements. You're making assumptions or guesses that are not congruent with the reality of the problem. 

-- 
-a

**Comment by akutz** [comment]
This has become a larger issue due to #684. It may no longer be possible to handle it automatically via a client option if the clients are ephemeral or lack interaction. It may only be possible to enable TLS on the server by default and via documentation require that clients set environment variables with encoded CA and client certs to successfully communicate with the server.

**Comment by akutz** [comment]
I am assigning this to @vladimirvivien as it relates to #584.

**Comment by vladimirvivien** [comment]
@akutz @cduchesne reading this issue I am not sure what the `ask` is for this.  Andrew mentions (and provides the link) of where this is already implemented.  I know we briefly talked about making configuration easier.  What are some usability features we want to ease out of this issue (if any)?

**Comment by vladimirvivien** [comment]
@akutz @cduchesne the one `ask` I can deduce is where/when to generate default certs.  Do we want the user to gen certs or have the installation process do it.  Is that the essence of this issue, are there other asks ?  If so, let's outline them here.

**Comment by akutz** [comment]
@vladimirvivien Please feel free to design and implement it as you see fit.

**Comment by vladimirvivien** [comment]
The following is a rough plan of what is to be implemented for this issue.  The goal is to allow RR to support TLS configurations that are aligned with what's supported in libstorage.

**Configure TLS from CLI**

Specify TLS mode as `insecured`:
```
$> rexray volume --tls insecured ls
``` 

Specify cert fingerprint:
```
$> rexray volume --tls "sha256:15:92:77:BE:6C:90...08:99" ls
``` 

Specify a trusted CA certs file:
```
$> rexray volume --tls-trusted-cert ca.crt ls
```

Specify client key/cert pair for authorization:
```
$> rexray volume --tls-client-cert client.crt --tls-client-key client.key ls
```

** Configure TLS from ENV variables **
The code will read TLS configurations from ENV variables as shown in the following examples:
```
$> LIBSTORAGE_CLIENT_TLS=insecure rexray volume ls
$> LIBSTORAGE_CLIENT_TLS="sha256:15:92:77:BE:6C:90...08:99" rexray volume ls
$> LIBSTORAGE_CLIENT_TLS_TRUSTEDCERTFILES=ca.crt rexray volume ls
$> LIBSTORAGE_CLIENT_TLS_CERTFILE=client.crt LIBSTORAGE_CLIENT_TLS_KEYFILE=client.key rexray volume ls
```

**Comment by codenrhoden** [comment]
So this would be my expectation for "default to TLS".

unless libstorage server TLS is explicitly disabled (which is a change from today, where it is disabled by default), when libstorage server is started up, it will look for a TLS cert. If it does not have one configured, it will generate a self-signed one based on the host name (this is the "snake oil" cert).

This means that by default, libStorage server will come up with TLS enabled. You *should* replace the TLS cert with a better one. You *can* disable it. But the default is that the server comes up with TLS enabled.

What say others?

**Comment by cduchesne** [comment]
@codenrhoden - i agree with your definition and would like to see libstorage operate in that way

**Comment by vladimirvivien** [comment]
Hi @codenrhoden @cduchesne 
Generating the certs automatically can be done.  But here are the issues with auto-gen
- Client doesn't validate server certs (they are auto created)
- Can do client authentication via cert
- The transport is encrypted, but insecure since there is no trusted CA provided

As such doing auto-generate will provided encryption but is still insecure.

Thoughts

**Comment by cduchesne** [comment]
"Default-to-TLS" from my definition is for encryption and not client auth. This can be done from the server-side only, combined with providing the generated certificate authority cert to each of the rexray clients, or by optionally enabling client.tls.insecure=true.

Although we also have support for client-side TLS for authentication, this is not part of what most think about when they think of TLS and should not be part of the auto-generating functionality.

**Comment by akutz** [comment]
I personally reject any proposal where the client has to operate in insecure mode. This is not security -- it's security theater and subject to MitM attacks. Unless the RR CLI is updated to prompt users to accept the remote server's certificate and remembers this for future connections then it's a terrible idea and should not be implemented. Also, the interactive nature of remembering the remote's certificate (unless it's signed by a known CA or the CA is configured on the client) means it won't work for non-interactive scenarios like REX-Ray's Docker volume endpoint modules.

**Comment by codenrhoden** [comment]
I agree with all your points, @vladimirvivien. And I know @akutz has strong opinions on it as well.

I thought pushing for this minimal form of encryption by default (using self-signed certs as the starting point) was what was being asked for. It is very true this is not high secure in any way.

Andrew made some points earlier in this issue about the need for client-side certs. I think that has been mitigated somewhat by the new token support. So you can have encrypted transport, but use the JWT tokens for auth if you like.

If we do autogenerate, @vladimirvivien, you had asked when/where? My suggestion would be:
when: at server start up (not package installation)
where: `libstorageHome + "tls"` or similar.

meaing, `/var/lib/libStorage/tls`. When the server starts up, the default config is to look for certs there. If they don't exist, they get generated.

**Comment by codenrhoden** [comment]
If users have internet access, we could just use `letsencrypt` for the CA. :)

**Comment by cduchesne** [comment]
@codenrhoden - that would only work if users were doing dns-based ssl validation and not also supporting ip-based; it would also be complex from a generation standpoint of making that even work ;-)

**Comment by akutz** [comment]
If we were able to assure publicly routed connections we could do a lot of neat things.

**Comment by cduchesne** [comment]
@akutz i agree with your rejection and was only calling out an already-available option that is there for someone to consume if they wanted to not authenticate the validity of the server's tls cert

**Comment by akutz** [comment]
IPSec. IPSec for everyone.

**Comment by codenrhoden** [comment]
I definitely meant it as a joke. 

Let me ask this question: Is there real value in preventing someone from capturing plain-text packets with WireShark? Is that security theater? People who want to prevent MITM attacks have to deploy a full CA chain, I get that.

We don't want to give people a false sense of security just because the transport is encrypted. But does that mean there is no value in having encrypted transport?

**Comment by akutz** [comment]
I'm saying if you're already limiting it to a private network and someone is snooping on that then how much harder is it to use Fiddler or Ethereal with a custom cert to listen to traffic?

**Comment by vladimirvivien** [comment]
@cduchesne @codenrhoden @akutz - We have to have a minimal set of assumptions about the person using RR.  One of them is that that person has the ability to create (self-signed or have) certificates and public key pairs.  So keep that in mind
- Token just handles authentication.  
- Using cert hash fingerprint basically just saves the user of not generating client keypairs.
- Auto-generating keypairs only gives us non-trusted encrypted transport (which means it's it's not buying you much in terms of security)

So, let's us come to an agreement here prior to EOD what we want to do here.

**Comment by cduchesne** [comment]
@akutz you would still need to become the proxy endpoint that the rexray client thinks is the rexray server to do any sort of mitm, as only the rexray server has the ssl key to decrypt said traffic

**Comment by akutz** [comment]
Yes, and it's a private network and I'm a disgruntled employee and place a host file somewhere or update DNS. It's not hard.

**Comment by cduchesne** [comment]
@akutz i am not disagreeing with that possibility, but there is much worse that could be done by a disgruntled employee just as easily if you want to go down that path ;-)

**Comment by akutz** [comment]
@vladimirvivien I think the RR CLI should prompt the user, like SSH, the first time it encounters an unknown cert presented by the server and ask them if they want to proceed. If so then the cert is saved and remembered as a valid endpoint.

**Comment by akutz** [comment]
Chris, for sure. I just don't wish to add to the list.

**Comment by cduchesne** [comment]
@vladimirvivien I don't think we should be mixing encryption and auth in the same discussion. The only auth that matters for this piece is that the server is who it says it is (authorizing the server's ssl cert as valid)

**Comment by vladimirvivien** [comment]
@cduchesne not mixing auth and secured transport at all.  I was pointing out that Tokens only handles auth.

**Comment by codenrhoden** [comment]
so, can we auto-generated a cert on the server side (if one wasn't provided), then ask that clients accept that cert on first contact, and persist that info to `/var/lib/libstorage/tls` ?  Yes, if someone is only using Docker, this presents an extra step (running a command to save the cert before the Docker volume plugin will work), but is that the compromise here?

**Comment by cduchesne** [comment]
@akutz i had forgotten about the implementation of client.tls supporting a an ssl fingerprint - perhaps we move this ssl fingerprint to a file in /var/lib/libstorage/tls that saves authorized fingerprints similar to ~/.ssh/known-hosts does

**Comment by akutz** [comment]
Chris, I hadn't forgotten that and I thought that's what I've been suggesting :)

**Comment by akutz** [comment]
And it would not be in etc but in $HOME/.libstorage/known_hosts as a user specific file.

**Comment by cduchesne** [comment]
@akutz my bad - i was implying that you reminded me and i agree with your suggestion

**Comment by akutz** [comment]
I know. I'm a dick even when people agree. :)

**Comment by cduchesne** [comment]
let me modify your statement
"I know. I'm a dick. :)"

**Comment by akutz** [comment]
FWIW, it would be technically set to insecure and there would be an additional validation that validated the thumbprint of the remote's peer cert. This is very similar to how I was going to use fingerprints on the server side per service prior to me deciding on tokens.

**Comment by akutz** [comment]
FTFY

"I know I'm a dick."

**Comment by codenrhoden** [comment]
>so, can we auto-generated a cert on the server side (if one wasn't provided), then ask that clients accept that cert on first contact, and persist that info to $HOME/.libstorage/tls ? Yes, if someone is only using Docker, this presents an extra step (running a command to save the cert before the Docker volume plugin will work), but is that the compromise here?

So is this an accurate statement of what we want to do?

**Comment by vladimirvivien** [comment]
I am summarizing:

- support auto-gen certs like so
 ```
# start server
$> rexray --tls-auto

# client  connects
$> rexray-client --tls-auto -h https://somehost
```
- client prompts for user to accept certs when unknown server is encountered
```
Attempting connect to unknown server, accept cert [y]? _
```
- Or, if keypairs and CA available, client can fully TLS
```
$> rexray-client --tls-cert-file <path/to/cert> --tls-key-file <path/to/keyfile> --tls-ca <path/to/ca>
```
@cduchesne @akutz @codenrhoden anything else ?

**Comment by codenrhoden** [comment]
I don't agree with the `--tls-auto` part. That should be the default, so should not be necessary.

**Comment by vladimirvivien** [comment]
Ok my next question, do we regenerate the TLS with every invocation of `rexray-client` or `rexray-controller` or save for later use ?

**Comment by cduchesne** [comment]
@vladimirvivien TLS certs created only once, on server. Fingerprint only asked about and saved to file when it isn't in file already, therefore only happens once.

**Comment by codenrhoden** [comment]
I would hope for re-use. I wouldn't want all clients to have to re-accept the cert every time the service is restarted.

**Comment by vladimirvivien** [comment]
@cduchesne where is fingerprint coming from ? I thought both sides are auto-creating certs ?

**Comment by cduchesne** [comment]
@vladimirvivien client-side TLS certs are for auth. The client generates the fingerprint based off of the public certificate the server sends to it upon connection.

**Comment by vladimirvivien** [comment]
@cduchesne ok.  So let me re-re-resummarize

- When `rexray` server starts it creates a keypair
- when `rexray-client` is invoked, it prompts to accept the server certs
- `rexray-client` then generates a fingerprint and stores that 
- Client then updates its config with `client.tls` to specify the fingerprint with `sha256` prefix
- anything else ?

**Comment by cduchesne** [comment]
@vladimirvivien yes however i believe the fingerprints should be stored in a different file and not the main config so it's easy to update remotely as well as store multiple fingerprints

**Comment by vladimirvivien** [comment]
@cduchesne Yes.  The fingerprint will be in its own file

**Comment by vladimirvivien** [comment]
ok

**Comment by codenrhoden** [comment]
I think that's pretty much it, @vladimirvivien. I would rewrite it like so:

- `rexray` server starts, generates new keypair if needed
- `rexray-client` connects to server, receipts server public cert, and calculates fingerprint
- `rexray-client` checks if fingerprint present in `libstorage.client.tls`
- `rexray-client` prompts user to accept fingerprint, if not already found
- if user accepts, fingerprint persisted to $HOME/.libstorage/tls/server_certs (or something)

When `rexray-client` starts up, the contents of that `tls/server_certs` file should be merged with `libstorage.client.tls`

**Comment by vladimirvivien** [comment]
@codenrhoden that sounds good.

**Comment by codenrhoden** [comment]
This is closed by #829
